src/Emogrifier/CssConcatenator.php

```diff
@@ -0,0 +1,122 @@
+     * - stdClass[] `ruleBlocks` - Array of rule blocks in order, where each element is an object with the following
+     *   properties:
+     *   - int[] `selectorsAsKeys` - Array whose keys are selectors for the rule block (values are of no significance);
+     *   - string `declarationsBlock` - The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     *
+     * @var stdClass[]
+     */
+    private $mediaRules = [];
+
+    /**
+     * Appends a declaration block to the CSS.
+     *
+     * @param string[] $selectors Array of selectors for the rule, e.g. ["ul", "ol", "p:first-child"].
+     * @param string $declarationsBlock The property declarations, e.g. "margin-top: 0.5em; padding: 0".
+     * @param string $media The media query for the rule, e.g. "@media screen and (max-width:639px)",
+     *                      or an empty string if none.
+     */
+    public function append(array $selectors, $declarationsBlock, $media = '')
+    {
+        $lastMediaRule = end($this->mediaRules);
+        if ($lastMediaRule !== false && $media === $lastMediaRule->media) {
+            $mediaRule = $lastMediaRule;
+        } else {
+            $mediaRule = (object)[
+                'media' => $media,
+                'ruleBlocks' => [],
+            ];
+            $this->mediaRules[] = $mediaRule;
+        }
+
+        $lastRuleBlock = end($mediaRule->ruleBlocks);
+        $selectorsAsKeys = array_flip($selectors);
+        if ($lastRuleBlock !== false && $declarationsBlock === $lastRuleBlock->declarationsBlock) {
+            $lastRuleBlock->selectorsAsKeys += $selectorsAsKeys;
+        } elseif ($lastRuleBlock !== false
+            && $this->hasEquivalentSelectors($selectorsAsKeys, $lastRuleBlock->selectorsAsKeys)
+        ) {
+            $lastRuleBlock->declarationsBlock
+                = rtrim(rtrim($lastRuleBlock->declarationsBlock), ';') . ';' . $declarationsBlock;
+        } else {
+            $mediaRule->ruleBlocks[] = (object)compact('selectorsAsKeys', 'declarationsBlock');
+        }
+    }
+
+    /**
+     * @return string
+     */
+    public function getCss()
+    {
+        $css = '';
+        foreach ($this->mediaRules as $mediaRule) {
+            if ($mediaRule->media !== '') {
+                $css .= $mediaRule->media . '{';
+            }
+            foreach ($mediaRule->ruleBlocks as $ruleBlock) {
+                $css .= implode(',', array_keys($ruleBlock->selectorsAsKeys))
+                    . '{' . $ruleBlock->declarationsBlock . '}';
+            }
+            if ($mediaRule->media !== '') {
+                $css .= '}';
+            }
+        }
+        return $css;
+    }
+
+    /**
+     * Tests if two sets of selectors are equivalent (i.e. the same selectors, possibly in a different order).
+     *
+     * @param int[] $selectorsAsKeys1 Array in which the selectors are the keys, and the values are of no significance.
+     * @param int[] $selectorsAsKeys2 Another such array.
```

Comment:
`int[]` for selectors does not seem to make sense here. Should be if `string[]` instead?

Reply:
It should probably be `mixed`.  It's whatever the keys happened to be in the original array of selectors passed, before `array_flip`, so it could be int or string.